### PR TITLE
i2c: add locking sem to sam0 driver

### LINKED ISSUE #42425: i2c: sam0 driver does not prevent simultaneous transactions
**Is your enhancement proposal related to a problem? Please describe.**
There are no protections against simultaneous i2c transactions colliding on the same bus. Identical to #29239.

**Describe the solution you'd like**
It seems like consensus from issues like #24511 is that a global i2c lock is the way to go, but I don't know what the status of that is.

**Proposed change**
Implement a locking semaphore for ```i2c_sam0_configure``` and ```i2c_sam0_transfer```.

### PR DESCRIPTION
**Pull request by maxmclau**
As is, the sam0 i2c driver has no mechanism to prevent simultaneous transactions.

Implemented a simple semaphore in i2c_sam0.c as seen in other platform driver.

Fixes #42425

### CODE REVIEW

File: drivers/i2c/i2c_sam0.c [hunk 1]
```diff
@@ -480,6 +484,8 @@ static int i2c_sam0_transfer(const struct device *dev, struct i2c_msg *msgs,
 			/* return the bus to idle */
 			i2c->CTRLB.bit.CMD = 3;
 
+			k_sem_give(&data->lock);
+
```
Comment (nandojve) [inline drivers/i2c/i2c_sam0.c:488]: Could this result in multiple "give" ?

**Review by nandojve — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_sam0.c (hunk 1 above)
Comment (teburd) [reply drivers/i2c/i2c_sam0.c:488]: This relates to the single exit path from the function being preferred, lock on the way in, unlock on the single path out

**Comment by teburd** [comment]
#43028 seems to solve the same issue, and the author has correctly setup a single exit path, please take a look to see if this solves your issue

**Comment by maxmclau** [comment]
Looks like a much better implementation, closing.

**Comment by attie-argentum** [comment]
Apologies! I should have looked for existing PRs...

**Pull request closed without merging**